Feature/dfa

natdevko opened the pull request:
Whole program in main, need review for values of a1, a2 problems
Probably polyfit needs to be corrected to use more accurate calculation

natdevko committed:
DFA - first attempt to polyfit with undefined reference to polyfit class

natdevko committed:
Whole program in main, need review for values of a1, a2 problems

Probably polyfit needs to be corrected to use more accurate calculation

bartek-krk commented on CMakeLists.txt:
```diff
@@ -8,7 +8,7 @@ set(CMAKE_AUTOUIC ON)
 # Comment the following line on Windows (uncomment on Mac)
 # DO NOT COMMIT THIS CHANGE
-set(CMAKE_PREFIX_PATH "/usr/local/Cellar/qt@5/5.15.7/lib/cmake/Qt5")
+set(CMAKE_PREFIX_PATH "C:\Qt\5.15.11\mingw81_64\lib\cmake\Qt5")
```
"`# DO NOT COMMIT THIS CHANGE`"
Względnie można ją zacommentować z opisem, że to działa na Windowsie (j.w.).

natdevko replied:
ktoś mi musi powiedzieć, jak naprawić ten plik, żeby go dobrze scommitować

bartek-krk commented on Modules/HRV_DFA/main.cpp:
```diff
@@ -0,0 +1,200 @@
+
+
+
+int main()
```
To nie powinno się tutaj znaleźć.

bartek-krk replied:
[ISSUE] - Jeżeli działanie algorytmu wymaga przeniesienia części jego logiki poza klasę zawierającą narusza to istotnie hermetyzację. Klasa powinna zawierać całą implementację wewnątrz i nie uzewnętrzniać szczegółów implementacji.

natdevko replied:
Algorytm będzie wymagał drobnej implementacji poza klasą. Jak będę mieć uporządkowany kod, to w mainie zawrę informację, które elementy muszą się znaleźć w aplikacji głównej (głównie chodzi o input i wybór przedziałów)

bartek-krk commented on Modules/HRV_DFA/lab3.m:
```diff
@@ -0,0 +1,98 @@
+clear
```
Pliki `*.m` nie powinny znajdować się w tym repozytorium.

natdevko replied:
poprawione

bartek-krk commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,145 @@
+
+ void sumDouble(vector<double>& diff2, double &sum)
+    {
+//        double temp = 0.0;
```
Zacommentowany kod.

natdevko replied:
poprawione

bartek-krk commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,145 @@
+        {
+            value = p.at(0)*x.at(i) + p.at(1);
+            y_temp.push_back(value);
+//            cout << "Test dzialania" << endl;
```
:)

natdevko replied:
poprawione, couty do debuggu mają komentarze

bartek-krk commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,145 @@
+
+using namespace std;
+
+void polyfit::calculateCoefficient(vector<int>& x)
```
Funkcje w tym pliku powinny być wywoływane w konstruktorze klasy, nie z zewnątrz.

natdevko replied:
z zewnątrz wywołanie funkcji jedynie dla funkcji publicznych

bartek-krk commented on Modules/HRV_DFA/DFA.h:
```diff
@@ -0,0 +1,92 @@
+
+using namespace std;
+
+class polyfit {
```
`polyfit` => `Polyfit`

natdevko replied:
zrobione

bartek-krk commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,145 @@
+            = (sum_y * sum_x_square - sum_x * sum_xy);
+        double denominator
+            = (N * sum_x_square - sum_x * sum_x);
+        constTerm = numerator / denominator;
```
Możliwe, że nie jest to zbyt realny scenariusz, ale co jeżeli `denominator == 0`?

natdevko replied:
denominator nigdy nie będzie równy 0, bo pracujemy na double'u z liczbami różnymi od 0. Ten algorytm i tak finalnie powinien być zmieniony na obliczanie macierzowe

natdevko replied:
zmienione i jw - nie ma opcji dzielenia przez 0

bartek-krk commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,145 @@
+            = (N * sum_xy - sum_x * sum_y);
+        double denominator
+            = (N * sum_x_square - sum_x * sum_x);
+        coeff = numerator / denominator;
```
`denominator == 0`

natdevko replied:
jak wcześniej, nie da się dzielić przez 0

bartek-krk commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,145 @@
+
+double polyfit::coefficient(vector<int>& x)
+    {
+        if (coeff == 0)
```
Dlaczego takie porównanie? Może on po prostu ma wynosić 0 dla niektórych przypadków

natdevko replied:
to już usunięte

bartek-krk commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,145 @@
+
+double polyfit::constant(vector<int>& x)
+    {
+        if (constTerm == 0)
```
Dlaczego takie porównanie? Może on po prostu ma wynosić 0 dla niektórych przypadków

natdevko replied:
usunięte

bartek-krk reviewed (changes requested):
1) Kod nie może posiadać żadnej funkcji `main`. Ta logika powinna zawierać się w ciele klasy. Jeżeli głównym zadaniem klasy `Polyfit` jest dopasowanie wielomianu do danych klasa MUSI posiadać dokładnie jedną metodę publiczną umożliwiającą uzyskanie wyników.
2) Nie commitujemy plików MATLAB - są tutaj zupełnie bezużyteczne
3) Nie commitujemy danych testowych - aplikacja będzie korzystać z plików dostarczonych przez użytkownika
4) Zacommentowany kod
5) Spodziewam się, że porównanie wartości pól klasy do zera to heurystyka polagajca na sprawdzeniu, czy pole zostało już zainicjalizowane. Co jeżeli naprawdę ma tam być 0?

MarcelPikula commented on Modules/HRV_DFA/DFA.h:
```diff
@@ -0,0 +1,92 @@
+    void calculateConstantTerm(vector<int>& x);
+ 
+    // Function that return the size of data set
+    int sizeOfData(vector<int>& x);
```
Czy to jest nam naprawdę potrzebne? Skoro chcemy uzyskać informację na temat wielkości wystarczy wywołać `vector.size()`.

natdevko replied:
nie jest, już wyrzuciłam, funkcja miała sens w zapożyczonym kodzie

MarcelPikula commented on Modules/HRV_DFA/DFA.h:
```diff
@@ -0,0 +1,92 @@
+        coeff = 0;
+        constTerm = 0;
+        sum_y = 0;
+        sum_y_square = 0;
```
Starajmy się używać camelCase'a. `sumYSquare`

natdevko replied:
poprawione

MarcelPikula commented on Modules/HRV_DFA/DFA.h:
```diff
@@ -0,0 +1,92 @@
+ 
+ 
+    // Function to take input from the dataset
+    void takeInput(vector<int>& x, vector<double>& yy);
```
Dlaczego zmienna nazywa się `yy`? Zmień proszę na pojedynczą literę

MarcelPikula commented on Modules/HRV_DFA/DFA.h:
```diff
@@ -0,0 +1,92 @@
+    void takeInput(vector<int>& x, vector<double>& yy);
+
+    // Function to return vector of 2 values - a and b
+    void returnP(vector<double>& p);
```
Nazwa metody ewidentnie wskazuje, że powinna coś zwracać, tymczasem mamy `void`

natdevko replied:
poprawione

MarcelPikula commented on Modules/HRV_DFA/DFA.h:
```diff
@@ -0,0 +1,92 @@
+
+    void diffVec(vector<double>& y_cut, vector<double>& yf, vector<double>& diff);
+
+    void pow2Vec(vector<double>& diff, vector<double>& diff2);
```
Wszystkie funkcje nie będące metodami klasy powinny znaleźć się w pliku `.cpp` w anonimowym namespace'e.

natdevko replied:
funkcje są już w klasach

MarcelPikula commented on Modules/HRV_DFA/DFA.h:
```diff
@@ -0,0 +1,92 @@
+
+    void sumDouble(vector<double>& diff2, double &sum);
+
+    void log10Vec(vector<double>& f, vector<double>& flog);
```
Nie widzę sensu implementacji tej samej funkcji dwukrotnie tylko z powodu innych typów trzymanych przez wektor. Spróbuj proszę podziałać z template.

natdevko replied:
template zrobiony

MarcelPikula commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,145 @@
+ void log10Vec(vector<double>& f, vector<double>& flog)
+    {
+        double temp = 0.0;
+        for( size_t i = 0; i < f.size(); i++ )
```
Wykorzystywanie pętli sumarycznie jest dość nie czytelne. Proszę spróbuj z `std::for_each`

natdevko replied:
zrobione pętlą auto it

MarcelPikula commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,145 @@
+    }
+
+
+ void sumDouble(vector<double>& diff2, double &sum)
```
Bezcelowa implementacja - mamy o wiele lepszą alternatywę: `std::reduce()`

natdevko replied:
zrobione for_each

MarcelPikula commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,145 @@
+    }
+
+
+ void sumVec(vector<double>& y, vector<double>& yk)
```
Funkcja swoją nazwą sugeruje, że na jej wyniku otrzymamy zsumowaną wartość wektora, jednak tak się nie dzieje. Zmień proszę nazwę na bardziej deskryptywną.

natdevko replied:
usunięta funkcja - pętla dodana w klasie

natdevko committed:
todo:
- more cleaning of code
- delete lines for debugging
- change algorithm to matrix

natdevko committed:
feature/DFA - todo:

todo:
- more cleaning of code
- delete lines for debugging
- change algorithm to matrix

bartek-krk reviewed (changes requested):
Ten PR wymaga jeszcze trochę pracy. W tym momencie projekt zawiera dwie funkcje main. Ten kod został prawdopodobnie napisany w osobnym edytorze i wklejony w to miejsce. Plik `CMakeLists.txt` został usunięty z projektu. Prawdopodobnie IDE wskazywało, że to błąd i usunięcie "rozwiązało" problem. Finalnie w aplikacji to nie zadziała... Nie możemy zmergować tego kodu dopóki nadmiarowa funkcja `main` nie zostanie usunięta, a plik nie wróci na swoje miejsce. Rozumiem sens klasy (I)`inputData`, ale jej instancję można stworzyć wewnątrz klasy (P)`polyfit`, co znacznie uprościłoby później użycie algorytmu. Komentarz odnośnie nazw zmiennych: nazwa `input` nie mówi zbyt wiele osobie, która tego nie napisała - nawet dla autora może być ona nieczytelna np. za miesiąc, chociaż doceniam komentarz w tym miejscu wyjaśniający jej znaczenie, ale takich przypadków jest więcej. Ostatni komentarz dotyczący zacommentowanego kodu: generalnie, rozumiem, że implementujemy ten moduł zanim warstwa DA jest dostępna, ale czy obecnie nieużywane rozwiązanie zostało przetestowane?

To rozwiązanie wymaga jeszcze trochę pracy, ale w stosunku do ostatniego review wygląda dużo lepiej.

Poniżej zamieszczam screenshot obecnego błędu który dostajemy.

<img width="756" alt="image" src="https://user-images.githubusercontent.com/62296439/205158706-77054bde-e360-41c7-9b0d-6cf431264aca.png">

natdevko commented:
> 

cmakelist usunęłam, bo mam tam zmiany, tak jak napisaliście w readme. Nie wiem, jak to poprawić, żeby nie robić błędu z początkowych commitów

kod pisałam w qt, ale jako osobny projekt, bo nie umiem stworzyć projektu, żeby wszystko było zbudowane jak w ostatecznej apce. Jeśli któryś z was jest w stanie poświęcić mi czas, to z pewnością zrobię to tak jak należy.

input mogę dać jako elementy klasy polyfit, po prostu uznałam, że dla czytelności powinny być osobno

w jaki sposób zakomentować zmienne, które muszą się pojawić w kodzie głównym? mam dokładnie opisać, skąd pochodzi ta zmienna i co zawiera?

kod na końcu .cpp był w domyśle do użycia przy innym algorytmie, ale prawdopodobnie z niego zrezygnuję, o ile uda mi się dopiąć bibliotekę gsl na własny użytek i przetestować jej działanie. Na razie nie usuwam, gdybym musiała robić operacje na macierzach. Przed 16.12 na pewno zniknie, jeśli nie będzie potrzebny

natdevko committed:
some code cleaning

TO DO:
- more cleaning of code
- delete lines for debugging
- add LS libraries for gls algorithm or use matrix algorithm

natdevko commented:
dodałam stare cmakelist, powinno być ok.

Reszta rzeczy jak w poprzednim komentarzu

MarcelPikula commented on Modules/HRV_DFA/DFA.h:
```diff
@@ -0,0 +1,149 @@
+#define ECG_ANALYZER_SAMPLE_H
+
+
+class inputData
```
Nazwa klasy z wielkiej litery

natdevko replied:
done

MarcelPikula commented on Modules/HRV_DFA/DFA.h:
```diff
@@ -0,0 +1,149 @@
+    vector <long double> y;
+    vector <long double> yk;
+    double xmean;
+    int length;
```
Nazwy pól klasy powinny przestrzegać konwencji `m_NameOfVariable`. Zastosuj to proszę do wszystkich przypadków swoich klas.

natdevko replied:
done

MarcelPikula commented on Modules/HRV_DFA/DFA.h:
```diff
@@ -0,0 +1,149 @@
+    public:
+    inputData()
+    {
+        xmean=0;
```
brakuje spacji między `=`

natdevko replied:
done

MarcelPikula commented on Modules/HRV_DFA/DFA.h:
```diff
@@ -0,0 +1,149 @@
+    vector <double> p;
+    vector <long double> y_est;
+    vector <long double> yf;
+    vector <int> xt;  //not used
```
Skoro `not used` to dlaczego jest tutaj?

natdevko replied:
usunięte

MarcelPikula commented on Modules/HRV_DFA/DFA.h:
```diff
@@ -0,0 +1,149 @@
+
+    void pow2Vec(vector<double>& diff, vector<double>& diff2);
+
+    double sumDouble(vector<double>& diff2, double &sum);
```
Tak jak rozmawialiśmy, nie potrzebujemy takiej funkcji. Mamy STL i `std:reduce()`

natdevko replied:
done

MarcelPikula commented on Modules/HRV_DFA/DFA.h:
```diff
@@ -0,0 +1,149 @@
+        tempx = 0;
+        tempy = 0;
+
+        sum = 0;
```
Pilnujmy wartości inicjalizujących - `int a = 0`, `float b = 1.f`, `double c = 2.0` i `long double d = 3.L`

natdevko replied:
usunięte

MarcelPikula commented on Modules/HRV_DFA/DFA.h:
```diff
@@ -0,0 +1,149 @@
+
+    // funkcja do liczenia wyznacznika macierzy, do algorytmu macierzowego
+    // not used
+    static double CalcDeterminant(vector<vector<double>> Matrix);
```
`not used` czyli kasujemy, racja?

natdevko replied:
usunięte

MarcelPikula commented on Modules/HRV_DFA/main.cpp:
```diff
@@ -0,0 +1,76 @@
+
+
+
+int main()
```
Ostateczna wersja **NIE MOŻE ZAWIERAĆ FUNKCJI MAIN**. Do debugowania spróbuj zrobić nowy projekt, który będzie nie zależny od tego co dzieje się na naszym repo. Cała funkcjonalność ma być przeniesiona do plików `DFA.cpp` i `DFA.h`

natdevko replied:
main wyrzucony

MarcelPikula commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,332 @@
+using namespace std;
+
+// creating array with natural numbers for dfa algorhitm
+vector<int> inputData::createNarray(int start, int end)
```
Ta implementacja nie jest optymalna. Proszę podziałaj z `std::iota()` np.:  

` std::vector<int> a(10);`
`std::iota(a.begin(), a.end(), 0);`

natdevko replied:
dopisuję do poprawy w następnym commicie

MarcelPikula commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,332 @@
+{
+    // this only for debugging
+    // delete before use in application
+    ifstream ifile ("C://Users//Natalia//Desktop//dadm projekt//ECG-Analyzer//Modules//HRV_DFA//nsr001.txt");
```
W ostatecznej wersji to musi zniknąć

natdevko replied:
tylko do debuggowania, w wersji do merga zniknie

MarcelPikula commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,332 @@
+
+    length = int(data.size());
+
+    for(int i = 0; i<length; i++)
```
`std::reduce`

natdevko replied:
done z accumulate

MarcelPikula commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,332 @@
+
+    xmean = xmean/length;
+
+    for( size_t i = 0; i < data.size(); i++ )
```
Pierwsza sprawa to: postarajmy się maksymalnie zmniejszyć liczbę operacji kopiowania i przypisania wartości wektorów. Dane wczytane mają znaczącą liczbę rekordów (setki tysięcy) więc każda operacja kopiowania znacząco odbije się nam na wydajności. Druga: nie używajmy pętli w takiej formie - `std::for_each` lub range-based loop. Zastosuj proszę te uwagi do całego swojego kodu

natdevko replied:
do poprawy w następnych commitach

MarcelPikula commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,332 @@
+        f.push_back(sq);
+
+        // for debugging
+        cout << "Test dzialania: " << j << endl;
```
Usuwamy

natdevko replied:
do debuggowania, w wersji do merga będzie usunięte

MarcelPikula commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,332 @@
+// y_est is y-value estimated from a and b values
+void Polyfit::calculateCoeff(vector<int>& x, vector<double>& p, vector<long double>& y_est)
+    {
+        a=0.0;
```
Brakuje spacji pomiędzy `=`

natdevko replied:
usunięte

MarcelPikula commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,332 @@
+    {
+        a=0.0;
+        b=0.0;
+        double N = x.size();
```
`x.size()` jako double? Zmień na `auto`

natdevko replied:
done w następnym commicie

natdevko committed:
improved algorithm

TO DO:
- proper descriptions of functions in .h
- reduce number of vectors
- check if calling results is properly made
- check if there is a chance of improving accuracy of calculating alfa1 (or is it from double precision)
- further cleaning of code

natdevko committed:
improved code readability and documentation comments
TO DO:
- check if documentation is made properly
- check "todo" in DFA.h
- reduce number of vectors
- check if alfa accuracy can be improved
- remember to update code in readInput before merge

natdevko committed:
improved code, documentation and alfa accuracy
TO DO:
- check if documentation is made properly
- check "todo" in DFA.h
- reduce number of vectors
- remember to update code in readInput before merge

natdevko committed:
improved optimalization, documentation and alfa accuracy
TO DO:
- check if documentation is made properly
- check "todo" in DFA.h
- check if further optimalization is possible (esp. readInput)
- remember to update code in readInput before merge
- check if GSL Library dependencies are prepared in main project

MarcelPikula commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,231 @@
+
+void Polyfit::readInput()
+{
+    // this only for debugging
```
Usuwamy elementy potrzebne do debugowania

natdevko replied:
Tak jak było wspomniane to zostało dlatego, że nie wiedziałam, czy jeszcze coś mam zmieniać w pozostałym kodzie, wtedy muszę sprawdzić, czy wszystko liczy się poprawnie

natdevko replied:
done

MarcelPikula commented on Modules/HRV_DFA/DFA.cpp:
```diff
@@ -0,0 +1,231 @@
+    int m_Length = int(m_Data.size());
+
+    // for proper application
+    // uncomment before merge
```
Proszę przetestuj działanie modułu z tym kodem. Jeżeli on coś wysypie to nie chcemy go mergować

natdevko replied:
Nie mam danych do tego modułu. W jaki sposób dostarczyć przykładowe dane?

natdevko replied:
done

MarcelPikula reviewed (changes requested):
Usuwamy niepotrzebne rzeczy, testujemy kod który jest obecnie zakomentarzowany - jeśli wszystko działa to po następnym commitcie będzie merge

natdevko committed:
commit for merge
TO DO:
- check if documentation is made properly
- check "todo" in DFA.h
    *todo - check if functions need to be safe to use (throw errors with bad data)
    *todo - check if any of returned value outside class needs to be std::shared_ptr<const std::vector<...>>
    *todo - check if libraries should be in cpp or h file
- check if further optimalization is possible (esp. readInput)
- check if GSL Library dependencies are prepared in main project

MarcelPikula reviewed (approved):
Nie obyło się bez drobnych problemów (odwoływanie przez referencje do wektorów zamiast kopiowania, zmiana argumentu jednej z metod z bool na enuma) ale ogólnie kod wygląda czysto. Zweryfikować komentarze TODO i jeśli wszystko jest w porządku i @bartek-krk da approve - mergujemy.
